Honkbots: The Clown's lil helpers.

Mercenaryblue opened the pull request:
[Changelogs]: https://github.com/tgstation/tgstation/wiki/Changelogs

:cl:
add: Use the Clown Stamp on some cardboard to begin... the honkbot!
add: These honkbots are just adorable, and totally not annoying. I swear! Honk!
add: You can even slot in a pAI! Just don't emag them... oh boy. oh no. oh geez.
soundadd: Honkbots now release an evil laugh when emagged.
/:cl:

[why]: Honkbots are silly little robots that have little purpose but to patrol and play a bike horn. They're good buddies when slotted with a pAI, a minor annoyance at best for most people, and a major annoyance to all when they become emagged. Other than that, its has been a long requested feature.

<a href="https://imgur.com/WxNNEw7"><img src="https://i.imgur.com/WxNNEw7.png" title="source: imgur.com" /></a>

Mercenaryblue committed:
Honkbots 0.9.0

Old code, needs testing.

Mercenaryblue committed:
Honkbots 0.9.1

Small Fixes - Part 1

Mercenaryblue committed:
Honkbot 0.9.1b

Fixes

Mercenaryblue committed:
Honkbots 0.9.2

Added sound cooldown, fixes.

Mercenaryblue committed:
Honkbots 0.9.5

Clown Box & Honkbot Assembly Code

Mercenaryblue committed:
0.9.5.a

oops

Mercenaryblue committed:
Honkbots 1.0.0

Final touches and fixes.
Clown Box finished.
Assembly functioning.
Honkbots tested and ready to go.
Its time.

FantasticFwoosh commented:
I can't say i really agree with how easily these are built, given you can order 50 cardboard sheets a pop from cargo.

- Actually requiring clown materials such as masks/bananskins etc to be built in the more conventional method from the robots crafting page would get my approval however.

praisenarsie commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,401 @@
+	var/arrest_type = 0
+	var/weaponscheck = 1
+
+/mob/living/simple_animal/bot/honkbot/New()
```
Initialize and . = ..()

praisenarsie commented:
@FantasticFwoosh If honkbots are abused, I'll just make a PR that bananium too is required to build them.

Kinda like gorillas, make them common then make them rare.

Mercenaryblue commented:
Bananium is a bit too much for something this low on the power creep level.
If they're abused, I'll remove the retaliation attack from normal behavior, making them easy kills.
Easy come, easy go.

Mercenaryblue committed:
Honkbots 1.0.1

Initialize and . = ..()

praisenarsie commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,404 @@
+	bot_core_type = /obj/machinery/bot_core/honkbot
+	window_id = "autosec"
+	window_name = "Honkomatic Bike Horn Unit v1.0.1"
+	allow_pai = 1 //Damn right we'll pAI these
```
this line is not needed, already allows pai in bot.dm

praisenarsie commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,404 @@
+
+	radio_key = /obj/item/device/encryptionkey/headset_service //doesn't have security key
+	radio_channel = "Service" //Service
+	bot_type = SEC_BOT
```
HONK_BOT ?

lzimann commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,43 @@
+
+/obj/item/stack/sheet/cardboard/attackby(obj/item/stamp/clown/I, mob/user, params)
+	if(!istype(I, /obj/item/stamp/clown))
+		..()
```
make this `return ..()`

praisenarsie commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,404 @@
+	dat += hack(user)
+	dat += showpai(user)
+	dat += text({"
+<TT><B>Honkomatic Bike Horn Unit v1.0 controls</B></TT><BR><BR>
```
make sure you update the version here too

Mercenaryblue replied:
Nice catch.

lzimann commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,43 @@
+		return
+	to_chat(user, "<span class='notice'>You stamp the cardboard! Its a clown box! Honk!</span>")
+	var/turf/T = get_turf(src)
+	var/obj/item/storage/box/clown/A = new /obj/item/storage/box/clown(T)
```
this whole line can be `var/obj/item/storage/box/clown/A = new`, also you can remote the `get_turf` since the item will be put on user's hand

jughu commented:
Can you add a special version that uses golden bike horns

lzimann commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,43 @@
+
+/obj/item/storage/box/clown/attackby(obj/item/I, mob/user, params)
+	if((istype(I, /obj/item/bodypart/l_arm/robot)) || (istype(I, /obj/item/bodypart/r_arm/robot)))
+		if(contents.len >= 1) //prevent accidently deleting contents
```
you only need `if(contents.len)` here

Mercenaryblue replied:
you do. without it, you'll create the assembly by deleting everything inside the box.

lzimann replied:
You might be confusing with just `contents`, what I said already implies it will be 1 or more(since 0 would not trigger this if and can't have list length < 0 )

Mercenaryblue replied:
I'll test it without, then.

Mercenaryblue replied:
Yeah, nah, without the .len - it won't even build an assembly, nor show the message.
We need it.

lordpidey replied:
We're not talking about removing the `.len`, we're talking about removing the `>=1`

Mercenaryblue replied:
Oh!!

lzimann commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,43 @@
+		qdel(I)
+		to_chat(user, "<span class='notice'>You add some wheels to the [src]! You've got an honkbot assembly now! Honk!</span>")
+		var/turf/T = get_turf(src)
+		var/obj/item/honkbot_assembly/A = new /obj/item/honkbot_assembly(T)
```
same thing as before

lzimann commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,43 @@
+		user.put_in_hands(A)
+		qdel(src)
+	else
+		..()
```
`return ..()`

lzimann commented on code/modules/mob/living/simple_animal/bot/construction.dm:
```diff
@@ -347,6 +347,44 @@
+
+/obj/item/honkbot_assembly/attackby(obj/item/I, mob/user, params)
+
+	if(isprox(I) && (build_step == 0))
```
make a define for the build steps(you can also add an #undef at the end of the file/after this attackby)

praisenarsie commented:
The sprites are outdated, you need to make them the new boxes rather than the old boxes.

lzimann commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,404 @@
+	var/target_lastloc //Loc of target when arrested.
+	var/last_found //There's a delay
+	var/threatlevel
+	var/declare_arrests = 0 // speak, you shall not, unless to Honk
```
use `FALSE` and `TRUE` for all of these vars

lzimann commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,404 @@
+/mob/living/simple_animal/bot/honkbot/Initialize()
+	. = ..()
+
+/mob/living/simple_animal/bot/honkbot/New()
```
remove this `New` and keep everything at initialize

praisenarsie commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,404 @@
+	bot_type = SEC_BOT
+	model = "Honkbot"
+	bot_core_type = /obj/machinery/bot_core/honkbot
+	window_id = "autosec"
```
autohonk

lzimann reviewed (dismissed):
there are still more stuff to review but i don't have the time now

praisenarsie commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,404 @@
+	var/arrest_type = 0
+	var/weaponscheck = 1
+
+/mob/living/simple_animal/bot/honkbot/Initialize()
```
already mentioned before on irc but replace New() with Initialize() and ..() with . = ..()

Mercenaryblue replied:
Done.

praisenarsie commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,404 @@
+	..()
+
+/obj/machinery/bot_core/honkbot
+	req_access = list(ACCESS_THEATRE)
```
```
req_one_access =list(ACCESS_THEATRE, ACCESS_ROBOTICS)
```

Mercenaryblue committed:
Honkbots 1.0.2

Fixes and Changes
====
Nerfed Brute/Thermal Resist
Buffed Health (to avoid 1-shot)
Added HONK_BOT define

Fox-McCloud commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,39 @@
+	//foldable = null
+
+/obj/item/stack/sheet/cardboard/attackby(obj/item/stamp/clown/I, mob/user, params)
+	if(!istype(I, /obj/item/stamp/clown))
```
It would make more sense to define if it *is* attacked by something, have the code there, then return, than to do this, in my opinion. That or `else return ..()` it.

Fox-McCloud commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,39 @@
+	illustration = null
+	//foldable = null
+
+/obj/item/stack/sheet/cardboard/attackby(obj/item/stamp/clown/I, mob/user, params)
```
It can be attacked by more than just this. This should be `obj/item/I`

Fox-McCloud commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,39 @@
+	to_chat(user, "<span class='notice'>You stamp the cardboard! Its a clown box! Honk!</span>")
+	var/obj/item/storage/box/clown/A = new
+	user.put_in_hands(A)
+	qdel(src)
```
It uses up an entire stack of cardboard no matter how much is in the stack? That's not good.

Also, you should always drop something before deleting it to avoid GC issues.

Cyberboss replied:
Destroy() takes care of removing it from inventories properly

Fox-McCloud replied:
I'm aware of that, but for whatever reason, it's not enough, and you explicitly have to drop the item before deleting it, from my experience.

GunHog commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,400 @@
+	window_id = "autohonk"
+	window_name = "Honkomatic Bike Horn Unit v1.0.2"
+	//allow_pai = 1 //Damn right we'll pAI these
+	data_hud_type = DATA_HUD_SECURITY_ADVANCED // show jobs
```
Use `DATA_HUD_SECURITY_BASIC` for jobs-only display.

ExcessiveUseOfCobblestone replied:
Why does it need to see jobs?

GunHog commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,400 @@
+
+/mob/living/simple_animal/bot/honkbot/set_custom_texts()
+
+	text_hack = "You overload [name]'s sound control system"
```
Oh come on, this is the FUNNY BOT! Give it some better texts! Maybe have it laugh evilly or something?

Mercenaryblue replied:
Oh yes, an evil clown laugh would be appropriate.

GunHog commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,400 @@
+
+/mob/living/simple_animal/bot/honkbot/proc/bike_horn() //use bike_horn
+	if (!emagged)
+		if(ckey == null) //check if a player is controlling
```
I recommend `if (!client)` here.

lordpidey commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,400 @@
+	else //emagged honkbots will spam short and memorable sounds.
+
+		if (ckey == null)
+			playsound(loc, pick('sound/items/bikehorn.ogg', 'sound/items/AirHorn2.ogg', 'sound/misc/sadtrombone.ogg', 'sound/items/AirHorn.ogg', 'sound/effects/reee.ogg', 'sound/effects/adminhelp.ogg', 'sound/items/WEEOO1.ogg', 'sound/voice/biamthelaw.ogg', 'sound/voice/bcreep.ogg','sound/magic/Fireball.ogg' ,'sound/effects/pray.ogg', 'sound/voice/hiss1.ogg','sound/machines/buzz-sigh.ogg', 'sound/machines ...
```
Take out the adminhelp.ogg, we do NOT want to train players to ignore that sound.

Mercenaryblue replied:
At 50% volume it was easily distinguishable from the actual admin bwoink.

Removing it now.

lordpidey commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,400 @@
+			// to be put at 100 volume: (Weeoo1, bcreep, blaw, hiss, flashbang )
+		else
+			if (!spam_flag)
+				playsound(loc, pick('sound/items/bikehorn.ogg', 'sound/items/AirHorn2.ogg', 'sound/misc/sadtrombone.ogg', 'sound/items/AirHorn.ogg', 'sound/effects/reee.ogg', 'sound/effects/adminhelp.ogg', 'sound/items/WEEOO1.ogg', 'sound/voice/biamthelaw.ogg', 'sound/voice/bcreep.ogg','sound/magic/Fireball.ogg' ,'sound/effects/pray.ogg', 'sound/voice/hiss1.ogg','sound/machines/buzz-sigh.ogg', 'sound/machine ...
```
Same here

lordpidey commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,400 @@
+				//target = old_target
+			else // you really don't want to hit an emagged honkbot
+				threatlevel = H.assess_threat(src)
+				threatlevel = 6 // will never let you go
```
This line makes the above line redundant.  There's no point in assessing the threat level if you're just gonna set it to 6 afterwards anyway.

lordpidey reviewed (changes requested):
Remove the adminhelp sounds, and one change about the threat detection

Bawhoppen commented:
You should use the old-fashioned sprite instead

Kerbin-Fiber commented:
You should make it that if someone finds the ban hammer, they can put it in the bot and THEN the bot can play the adminhelp.ogg

praisenarsie commented:
@Kerbin-Fiber No

Cyberboss commented on code/modules/mob/living/simple_animal/bot/construction.dm:
```diff
@@ -347,6 +348,41 @@
+	icon = 'icons/mob/aibots.dmi'
+	icon_state = "honkbot_arm"
+	item_state = "honkbot_arm"
+	var/build_step = 0
```
use defines for these

Mercenaryblue replied:
Took a while, but done.

Mercenaryblue commented:
Nah. If they want the bwoink out, we're removing it.

lordpidey commented:
Note that I'm not a maintainer, so I'm just marking an observation, I have no real authority.

Mercenaryblue commented:
Nevertheless, it is helping.

Mercenaryblue committed:
Honkbots 1.0.3

Removed the cardboard assembly from death gibs.
Admin Bwoink removed from emagged soundtrack.
Honkbot now release an evil laugh when emagged.
Fixed strange behaviors on emagged level 1
Added a check against stamping multiple cardboard sheets.
Stamping a Clown Box now plays the bike horn, once.
Clown Boxes no longer get automatically placed in hands.
Various other fixes.

lordpidey commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -615,14 +615,19 @@ obj/item/storage/box/clown
+/obj/item/stack/sheet/cardboard/attackby(obj/item/I, mob/user, params)
 	if(!istype(I, /obj/item/stamp/clown))
 		return ..()
+	if (src.amount >= 2)
```
change to `if(get_amount() >= 2)`

lordpidey commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -615,14 +615,19 @@ obj/item/storage/box/clown
+	playsound(loc, 'sound/items/bikehorn.ogg', 50, 1, -1)
+	var/turf/T = get_turf(src)
+	new/obj/item/storage/box/clown(T)
 	qdel(src)
```
Don't `qdel(src)`, instead do `use(2)`

Mercenaryblue replied:
use(2) won't delete the sheet, allowing to spawn infinite boxes from one sheet.

lordpidey reviewed (changes requested):
Couple of changes to the cardboard sheet code

Mercenaryblue committed:
Honkbots 1.0.3a

Mercenaryblue committed:
Honkbots 1.0.3.b

Fixes.

YakumoChen commented:
Instead of just a stamp you could make it require a banana

YakumoChen commented:
>plays sadtrombone if you trip on it

Laffo :+1:

Cyberboss commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,43 @@
+	if(!user.temporarilyRemoveItemFromInventory(src))
+		return
+	to_chat(user, "<span class='notice'>You stamp the cardboard! Its a clown box! Honk!</span>")
+	playsound(loc, 'sound/items/bikehorn.ogg', 50, 1, -1)
```
playsound(src, ...

Cyberboss commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,43 @@
+	to_chat(user, "<span class='notice'>You stamp the cardboard! Its a clown box! Honk!</span>")
+	playsound(loc, 'sound/items/bikehorn.ogg', 50, 1, -1)
+	var/turf/T = get_turf(src)
+	new/obj/item/storage/box/clown(T)
```
create it at drop_location()

Cyberboss commented on code/modules/mob/living/simple_animal/bot/construction.dm:
```diff
@@ -347,6 +348,41 @@
+			return
+		to_chat(user, "<span class='notice'>You add the bike horn to [src]! Honk!</span>")
+		var/turf/T = get_turf(src)
+		var/mob/living/simple_animal/bot/honkbot/S = new /mob/living/simple_animal/bot/honkbot(T)
```
var/mob/living/simple_animal/bot/honkbot/S = new(drop_location())

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	var/fcheck = 1 //And armed people
+	var/check_records = 1 //Doesn't care about criminals
+	var/arrest_type = 0
+	var/weaponscheck = 1
```
TRUE?

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	var/declare_arrests = 0 // speak, you shall not, unless to Honk
+	var/idcheck = 1 //Chases unknowns
+	var/fcheck = 1 //And armed people
+	var/check_records = 1 //Doesn't care about criminals
```
TRUE?

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	var/threatlevel = FALSE
+	var/declare_arrests = 0 // speak, you shall not, unless to Honk
+	var/idcheck = 1 //Chases unknowns
+	var/fcheck = 1 //And armed people
```
TRUE?

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	var/last_found = FALSE	//There's a delay
+	var/threatlevel = FALSE
+	var/declare_arrests = 0 // speak, you shall not, unless to Honk
+	var/idcheck = 1 //Chases unknowns
```
TRUE?

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	var/target_lastloc = FALSE	//Loc of target when arrested.
+	var/last_found = FALSE	//There's a delay
+	var/threatlevel = FALSE
+	var/declare_arrests = 0 // speak, you shall not, unless to Honk
```
FALSE?

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+
+	var/honksound = 'sound/items/bikehorn.ogg' //customizable sound
+	var/goldenbikehorn = 0 //placeholder
+	var/spam_flag = 0
```
FALSE?

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	data_hud_type = DATA_HUD_SECURITY_BASIC // show jobs
+
+	var/honksound = 'sound/items/bikehorn.ogg' //customizable sound
+	var/goldenbikehorn = 0 //placeholder
```
If its not used now, remove it

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	..()
+	target = null
+	oldtarget_name = null
+	anchored = 0
```
FALSE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	icon_state = "honkbot[on]"
+	auto_patrol = 1
+	spawn(3)
+		var/datum/job/clown/J = new/datum/job/clown
```
There's no need for this spawn

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+/mob/living/simple_animal/bot/honkbot/Initialize()
+	. = ..()
+	icon_state = "honkbot[on]"
+	auto_patrol = 1
```
TRUE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+
+/mob/living/simple_animal/bot/honkbot/Topic(href, href_list)
+	if(..())
+		return 1
```
No reason for this topic

Mercenaryblue replied:
I think it was a failsafe.

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+/mob/living/simple_animal/bot/honkbot/attack_hand(mob/living/carbon/human/H)
+	if(H.a_intent == "harm")
+		retaliate(H)
+		spawn(5)
```
Use addtimer

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	if(H.a_intent == "harm")
+		retaliate(H)
+		spawn(5)
+			playsound(loc, 'sound/machines/buzz-sigh.ogg', 50, 1, -1)
```
playsound(src, ...

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+
+
+/mob/living/simple_animal/bot/honkbot/attackby(obj/item/W, mob/user, params)
+	..()
```
call ..() last

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+		return
+	if(!istype(W, /obj/item/screwdriver) && (W.force) && (!target) && (W.damtype != STAMINA) ) // Check for welding tool to fix #2432.
+		retaliate(user)
+		spawn(5)
```
use addtimer

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	if(!istype(W, /obj/item/screwdriver) && (W.force) && (!target) && (W.damtype != STAMINA) ) // Check for welding tool to fix #2432.
+		retaliate(user)
+		spawn(5)
+			playsound(loc, 'sound/machines/buzz-sigh.ogg', 50, 1, -1)
```
playsound(src, ...

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+Behaviour controls are [locked ? "locked" : "unlocked"]<BR>
+Maintenance panel panel is [open ? "opened" : "closed"]"},
+
+"<A href='?src=\ref[src];power=1'>[on ? "On" : "Off"]</A>" )
```
power=[TRUE]

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+/mob/living/simple_animal/bot/honkbot/bullet_act(obj/item/projectile/Proj)
+	if(istype(Proj ,/obj/item/projectile/beam)||istype(Proj,/obj/item/projectile/bullet))
+		if((Proj.damage_type == BURN) || (Proj.damage_type == BRUTE))
+			if(!Proj.nodamage && Proj.damage < src.health)
```
This could be 1 if statement

Cyberboss replied:
remove src

Mercenaryblue replied:
Done.

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+
+/mob/living/simple_animal/bot/honkbot/hitby(atom/movable/AM, skipcatch = 0, hitpush = 1, blocked = 0)
+	if(istype(AM, /obj/item))
+		playsound(loc, honksound, 50, 1, -1)
```
playsound(src,

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	if(istype(AM, /obj/item))
+		playsound(loc, honksound, 50, 1, -1)
+		var/obj/item/I = AM
+		if(I.throwforce < src.health && I.thrownby && (istype(I.thrownby, /mob/living/carbon/human)))
```
remove src

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+			playsound(loc, honksound, 50, 1, -1)
+		else
+			if (!spam_flag)
+				playsound(loc, honksound, 50, 1, -1)
```
playsound(src,

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+/mob/living/simple_animal/bot/honkbot/proc/bike_horn() //use bike_horn
+	if (emagged <= 1)
+		if(!client) //check if a player is controlling
+			playsound(loc, honksound, 50, 1, -1)
```
playsound(src,

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+		icon_state = "honkbot-c"
+		spawn(5)
+			icon_state = "honkbot[on]"
+		spawn(cooldowntimehorn)
```
Use timers

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	else if (emagged == 2) //emagged honkbots will spam short and memorable sounds.
+
+		if (!client)
+			playsound(loc, pick('sound/items/bikehorn.ogg', 'sound/items/AirHorn2.ogg', 'sound/misc/sadtrombone.ogg', 'sound/items/AirHorn.ogg', 'sound/effects/reee.ogg',  'sound/items/WEEOO1.ogg', 'sound/voice/biamthelaw.ogg', 'sound/voice/bcreep.ogg','sound/magic/Fireball.ogg' ,'sound/effects/pray.ogg', 'sound/voice/hiss1.ogg','sound/machines/buzz-sigh.ogg', 'sound/machines/ping.ogg', 'sound/weapons/fla ...
```
playsound(src,

Cyberboss replied:
use get_sfx

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+			// to be put at 100 volume: (Weeoo1, bcreep, blaw, hiss, flashbang )
+		else
+			if (!spam_flag)
+				playsound(loc, pick('sound/items/bikehorn.ogg', 'sound/items/AirHorn2.ogg', 'sound/misc/sadtrombone.ogg', 'sound/items/AirHorn.ogg', 'sound/effects/reee.ogg', 'sound/items/WEEOO1.ogg', 'sound/voice/biamthelaw.ogg', 'sound/voice/bcreep.ogg','sound/magic/Fireball.ogg' ,'sound/effects/pray.ogg', 'sound/voice/hiss1.ogg','sound/machines/buzz-sigh.ogg', 'sound/machines/ping.ogg', 'sound/weapons/fla ...
```
playsound(src,

Cyberboss replied:
Use get_sfx

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+		else
+			if (!spam_flag)
+				playsound(loc, pick('sound/items/bikehorn.ogg', 'sound/items/AirHorn2.ogg', 'sound/misc/sadtrombone.ogg', 'sound/items/AirHorn.ogg', 'sound/effects/reee.ogg', 'sound/items/WEEOO1.ogg', 'sound/voice/biamthelaw.ogg', 'sound/voice/bcreep.ogg','sound/magic/Fireball.ogg' ,'sound/effects/pray.ogg', 'sound/voice/hiss1.ogg','sound/machines/buzz-sigh.ogg', 'sound/machines/ping.ogg', 'sound/weapons/fla ...
+				spam_flag = 1 // prevent spam
```
TRUE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+		icon_state = "honkbot-e"
+		spawn(30) // keep flashing
+			icon_state = "honkbot[on]"
+		spawn(cooldowntimehorn)
```
Use timers

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+		spawn(30) // keep flashing
+			icon_state = "honkbot[on]"
+		spawn(cooldowntimehorn)
+			spam_flag = 0
```
FALSE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+
+/mob/living/simple_animal/bot/honkbot/proc/honk_attack(mob/living/carbon/C) // horn attack
+	if (!client) //check if a player is controlling
+		playsound(loc, honksound, 50, 1, -1)
```
playsound(src,

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+		playsound(loc, honksound, 50, 1, -1)
+	else
+		if(spam_flag == 0)
+			playsound(loc, honksound, 50, 1, -1)
```
playsound(src,

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	if (!client) //check if a player is controlling
+		playsound(loc, honksound, 50, 1, -1)
+	else
+		if(spam_flag == 0)
```
if(!spam_flag)

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	else
+		if(spam_flag == 0)
+			playsound(loc, honksound, 50, 1, -1)
+			spam_flag = 1 // prevent spam
```
TRUE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	icon_state = "honkbot-c"
+	spawn(5)
+		icon_state = "honkbot[on]"
+	spawn(cooldowntimehorn)
```
Use timers

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+
+/mob/living/simple_animal/bot/honkbot/proc/stun_attack(mob/living/carbon/C) // airhorn stun
+	if (!client)
+		playsound(loc, 'sound/items/AirHorn.ogg', 100, 1, -1) //HOOOOOOOOOOOOONK!!
```
playsound(src,

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	spawn(5)
+		icon_state = "honkbot[on]"
+	spawn(cooldowntimehorn)
+		spam_flag = 0
```
FALSE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	if (!client)
+		playsound(loc, 'sound/items/AirHorn.ogg', 100, 1, -1) //HOOOOOOOOOOOOONK!!
+	else
+		if(spam_flag == 0)
```
if(!spam_flag)

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+		if(spam_flag == 0)
+			playsound(loc, 'sound/items/AirHorn.ogg', 100, 1, -1) //HEEEEEEEEEEEENK!!
+	icon_state = "honkbot-c"
+	spawn(5)
```
Use addtimer

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+		playsound(loc, 'sound/items/AirHorn.ogg', 100, 1, -1) //HOOOOOOOOOOOOONK!!
+	else
+		if(spam_flag == 0)
+			playsound(loc, 'sound/items/AirHorn.ogg', 100, 1, -1) //HEEEEEEEEEEEENK!!
```
playsound(src,

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	icon_state = "honkbot-c"
+	spawn(5)
+		icon_state = "honkbot[on]"
+	if(spam_flag == 0)
```
if(!spam_flag)

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+			C.Knockdown(80)
+			var/mob/living/carbon/human/H = C
+			if (ckey != null) //don't alter AI behavior
+				spam_flag = 1
```
TRUE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+			C.Jitter(50)
+			C.Knockdown(80)
+			var/mob/living/carbon/human/H = C
+			if (ckey != null) //don't alter AI behavior
```
if(ckey)

Mercenaryblue replied:
if you ghost with if(ckey), it just makes it braindead.
With (ckey != null), they return to their AI behavior.

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+				//target = old_target
+			else // you really don't want to hit an emagged honkbot
+				threatlevel = 6 // will never let you go
+			spawn(cooldowntime)
```
addtimer

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+			else // you really don't want to hit an emagged honkbot
+				threatlevel = 6 // will never let you go
+			spawn(cooldowntime)
+				spam_flag = 0
```
FALSE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+			C.Knockdown(80)
+
+	add_logs(src,C,"honked")
+	spawn(cooldowntime)
```
addtimer

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+
+	add_logs(src,C,"honked")
+	spawn(cooldowntime)
+		spam_flag = 0
```
FALSE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	return
+
+/mob/living/simple_animal/bot/honkbot/proc/back_to_idle()
+	anchored = 0
```
FALSE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	last_found = world.time
+	frustration = 0
+	spawn(0)
+		handle_automated_action() //ensure bot quickly responds
```
INVOKE_ASYNC

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	frustration = 0
+	mode = BOT_HUNT
+	spawn(0)
+		handle_automated_action() //ensure bot quickly responds
```
INVOKE_ASYNC

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+
+/mob/living/simple_animal/bot/honkbot/proc/back_to_hunt()
+	anchored = 0
+	frustration = 0
```
FALSE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+		handle_automated_action() //ensure bot quickly responds
+
+/mob/living/simple_animal/bot/honkbot/proc/back_to_hunt()
+	anchored = 0
```
FALSE

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+	mode = BOT_IDLE
+	target = null
+	last_found = world.time
+	frustration = 0
```
FALSE

Mercenaryblue replied:
Frustration isn't a TRUE or FALSE variable. It builds up.  See BOT_HUNT.

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,398 @@
+
+
+/mob/living/simple_animal/bot/honkbot/proc/look_for_perp()
+	anchored = 0
```
FALSE

Cyberboss reviewed (changes requested):
Okay, there's tons more. Why does this look like it was copypasted from secbots?

Mercenaryblue committed:
Honkbots 1.0.4

*Added an emote ping upon assembly creation.
*Enabled either Theatre or Robotics for access. ((fixing needing both))
*Honkbots are now more forgiving after being hit.
**Airhorn stun_attack now deals slight ear damage, for about 5 seconds.
**Cardboard no longer drops upon death. Robot arms have 50% chance.
**Using a new get_sfx() list when emagged.
**Optimization and other minor fixes.

Mercenaryblue committed:
Honkbots 1.0.4.a

Missed it.

GunHog commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -151,89 +161,83 @@ Maintenance panel panel is [open ? "opened" : "closed"]"},
 			var/mob/living/carbon/human/H = C
-			if (ckey != null) //don't alter AI behavior
-				spam_flag = 1
+			if(ckey != null) //used over (!ckey) to restore botAI when ghosting.
```
I do not think you should ever be checking for a ckey for honkbots, just if there is a client or not. If the player in control of the bot disconnects, the bot will still have a ckey. If you use client checks instead, the bot can resume normal AI behavior until the player reconnects. 

This is, of course, only a recommendation.

Mercenaryblue replied:
Using (!client) makes perfect sense, actually.

Mercenaryblue committed:
Honkbots 1.0.4.b

use(1) and (client)

Mercenaryblue commented:
Right. Tell me if I missed anything.

lzimann commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,42 @@
+	icon_state = "clownbox"
+	illustration = null
+
+/obj/item/stack/sheet/cardboard/attackby(obj/item/I, mob/user, params)
```
move this proc to be with the cardboard definition([here](https://github.com/tgstation/tgstation/blob/master/code/game/objects/items/stacks/sheets/sheet_types.dm))

Mercenaryblue replied:
Done.

Mercenaryblue committed:
Honkbots 1.0.4.c

Moved proc

GunHog commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -252,6 +252,19 @@ GLOBAL_LIST_INIT(cardboard_recipes, list ( \
+	if(!istype(I, /obj/item/stamp/clown))
+		return ..()
+	if(get_amount() >= 2)
+		to_chat(user, "<span class ='notice'>You only need one cardboard sheet for this.</span>")
```
The `use()` proc handles this kind of thing with sheets. Use it in an if() statement. `use(5)` would use five items in a stack, deleting the stack if exactly 5 were remaining, for example. You could do something similar here.

Mercenaryblue replied:
Yeah, but I've found a bug if you held the cardboard in your hands while having more than the use().
It use all the stack to make a box, and when you unfold, it returns to the stack number.

Mercenaryblue replied:
I'll give it a try.

Mercenaryblue replied:
Its a bit more complex than anticipated, give me some time.

Mercenaryblue replied:
That took a while, but I did it.

Mercenaryblue committed:
Honkbots 1.0.5

*Code Optimization
*Renaming w/ Pen (how could I even forget this)
*Assembly Defines

I'm afraid I'm simply out of my depth with the suggested use(1) rework.

Mercenaryblue committed:
Honkbots 1.0.5.c

Reworked cardboard stamping.

It now works in the hands and on the ground, either solo or in stacks, and prevent itself being used in backpacks.

lordpidey commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -255,15 +255,17 @@ GLOBAL_LIST_INIT(cardboard_recipes, list ( \
+	if ((src.amount > 1) && (user.is_holding(src))) //prevent a bug when holding stack
+		amount-- //you also keep the stack in your hand
+	else
+		use(1)
```
Just replace this whole proc with
```
if(istype(I, /obj/item/stamp/clown) && use (1))
    to_chat(user, "<span class='notice'>You stamp the cardboard! Its a clown box! Honk!</span>")
    new/obj/item/storage/box/clown(drop_location())
else
    . = ..()
```

Mercenaryblue replied:
Aight. Just gotta fix a couple bugs first, and i'll commit it.

Mercenaryblue replied:
use(1) is still returning too soon - it won't proc anything with a single sheet.
We need to move the sound and text message above the if(use).
It also prevents the drop_location() as it can't read where it is, since the source gets deleted.

Unless you can find a way to memorize the location on another object, we're gonna have to use the turf as a placeholder. Its not optimal, but it'll work.

lordpidey replied:
Aright, you're right, I forgot QDEL  immediately changes it's loc, give me a few minutes I'll fix it

Mercenaryblue replied:
I was about to push this: 
<script src="https://gist.github.com/Mercenaryblue/db910be0f6285e068819277114bacaab.js"></script>

lordpidey replied:
```
if(istype(I, /obj/item/stamp/clown))
    var/atom/droploc = drop_location()
    if(use(1))
        playsound(I, 'sound/items/bikehorn.ogg', 50, 1, -1)
        to_chat(user, "<span class='notice'>You stamp the cardboard! Its a clown box! Honk!</span>")
        new/obj/item/storage/box/clown(drop_location())
else
    . = ..()
```
    
This should work.  I've changed it to make the horn honk rather than the box, (Edge case involving TK mutation)

Mercenaryblue replied:
Onward to testing!

Mercenaryblue committed:
Honkbots 1.0.5.d

Final

Mercenaryblue committed:
Honkbots 1.0.5.e

comment

lordpidey commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -253,19 +253,15 @@ GLOBAL_LIST_INIT(cardboard_recipes, list ( \
+			playsound(I, 'sound/items/bikehorn.ogg', 50, 1, -1)
+			to_chat(user, "<span class='notice'>You stamp the cardboard! Its a clown box! Honk!</span>")
+			if (amount == 0) new/obj/item/storage/box/clown(droploc) //bugfix
+			else new/obj/item/storage/box/clown(drop_location())
```
I don't understand the purpose of this if and else statement

What happens if instead of this whole if statement, you simply have `new/obj/item/storage/box/clown(droploc) `?

Mercenaryblue replied:
tl;dr? drop_location() doesn't work with single stack.

If we use box/clown(droploc) for both, they'd both act the same.
We could save a line by removing 262, I guess.

Mercenaryblue committed:
1.0.5.f

Optimization

Mercenaryblue committed:
Honkbots 1.0.5.g

Missed requested stuff

Cyberboss commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -252,6 +252,16 @@ GLOBAL_LIST_INIT(cardboard_recipes, list ( \
+		if(use(1))
+			playsound(I, 'sound/items/bikehorn.ogg', 50, 1, -1)
+			to_chat(user, "<span class='notice'>You stamp the cardboard! Its a clown box! Honk!</span>")
+			if (amount >= 0) new/obj/item/storage/box/clown(droploc) //bugfix
```
multiline this if please

Cyberboss commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -252,6 +252,16 @@ GLOBAL_LIST_INIT(cardboard_recipes, list ( \
+			playsound(I, 'sound/items/bikehorn.ogg', 50, 1, -1)
+			to_chat(user, "<span class='notice'>You stamp the cardboard! Its a clown box! Honk!</span>")
+			if (amount >= 0) new/obj/item/storage/box/clown(droploc) //bugfix
+	else . = ..()
```
ditto

Cyberboss commented on code/game/objects/items/storage/boxes.dm:
```diff
@@ -609,6 +609,29 @@
+		to_chat(user, "<span class='notice'>You add some wheels to the [src]! You've got an honkbot assembly now! Honk!</span>")
+		var/obj/item/honkbot_assembly/A = new
+		user.put_in_hands(A)
+		qdel(src)
```
qdel(src0 before put_in_hands

Cyberboss commented on code/modules/mob/living/simple_animal/bot/construction.dm:
```diff
@@ -1,3 +1,4 @@
+#define isbuild_step(A) (build_step == A)
```
I'd rather you didn't user this define

Mercenaryblue replied:
Should I move it to helpers?

Mercenaryblue replied:
You want me to revert or move the definition?

Cyberboss commented on code/modules/mob/living/simple_animal/bot/construction.dm:
```diff
@@ -347,6 +348,47 @@
+		if(!user.temporarilyRemoveItemFromInventory(I))
+			return
+		build_step++
+		to_chat(user, "<span class='notice'>You add the prox sensor to [src]!</span>")
```
You add [I] to [src]

Cyberboss commented on code/modules/mob/living/simple_animal/bot/construction.dm:
```diff
@@ -347,6 +348,47 @@
+	else if(istype(I, /obj/item/bikehorn) && isbuild_step(ASSEMBLY_SECOND_STEP))
+		if(!user.temporarilyRemoveItemFromInventory(I))
+			return
+		to_chat(user, "<span class='notice'>You add the bike horn to [src]! Honk!</span>")
```
again

Cyberboss commented on code/modules/mob/living/simple_animal/bot/construction.dm:
```diff
@@ -347,6 +348,47 @@
+		if(!user.temporarilyRemoveItemFromInventory(I))
+			return
+		to_chat(user, "<span class='notice'>You add the bike horn to [src]! Honk!</span>")
+		var/mob/living/simple_animal/bot/honkbot/S = new(drop_location())
```
do S.loc since this is a transofrmation

Mercenaryblue replied:
Was I meant to replace this line, or simply add S.loc after?

Mercenaryblue replied:
new(S.loc) will produce an self-referencing infinite loop. 
I can either use new(loc), or use S.loc after generating the item.

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,388 @@
+
+/mob/living/simple_animal/bot/honkbot/turn_off()
+	..()
+	icon_state = "honkbot[on]"
```
icon_state changes in update_icon

Cyberboss commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,388 @@
+
+/mob/living/simple_animal/bot/honkbot/turn_on()
+	..()
+	icon_state = "honkbot[on]"
```
icon_state changes in update_icon

Mercenaryblue committed:
Honkbots 1.0.6

Requested changes.

Mercenaryblue committed:
Honkbots 1.0.6a

Oops.

Mercenaryblue committed:
Honkbots 1.0.6.c

Bugfix - Prevent building in backpacks.
Make sure they spawn on turf.

AnturK commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,380 @@
+						honk_attack(target)
+					else
+						if(threatlevel >= 6)
+							spawn(0)
```
waitfor =  FALSE on stun attack and moving the anchor resets there should be better.

Mercenaryblue replied:
to replace the spawn(0), right? On it.

AnturK commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,380 @@
+		if(threatlevel <= 3)
+			if(C in view(4,src)) //keep the range short for patrolling
+				if(!spam_flag)
+					spam_flag = TRUE
```
spam flag setting and resetting should be handled in bike_horn, if you need to skip that somewhere add a parameter.

Mercenaryblue replied:
Fixed.

Mercenaryblue commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,380 @@
+
+/mob/living/simple_animal/bot/honkbot/proc/bike_horn() //use bike_horn
+	if (emagged <= 1)
+		if(!client) //check if a player is controlling
```
Emagged 0 is the normal.
Emagged 1 is when you bypass the controls
Emagged 2 is when you short-circuit the whole thing.

They need to be acting normal until emagged 2.

Mercenaryblue replied:
if(!client) is used to make sure the honkbots don't go braindead if the ckey ghost.
I forgot about that one.

Mercenaryblue committed:
Honkbots 1.0.7

one less spawn(0)
flag optimization
else return ..()

Mercenaryblue committed:
Honkbots 1.0.7.a

Get rids of a lot of client checks at the cost of automatic AI retoration.

AnturK commented on code/modules/mob/living/simple_animal/bot/honkbot.dm:
```diff
@@ -0,0 +1,360 @@
+	return	dat
+
+/mob/living/simple_animal/bot/honkbot/proc/retaliate(mob/living/carbon/human/H)
+	threatlevel = H.assess_threat(src)
```
Assess threat needs flags variable not the user. Same other places where you use it.

Mercenaryblue committed:
Honkbots 1.0.7.b

Added judgment_criteria()
changed assess_threat to use judgment

Mercenaryblue committed:
Honkbots 1.0.7.c

final = NONE
